refine QuickNodeSetup/fsnNode.sh

cross-chain opened the pull request:
1. fix bug #58, `ubuntu 19.10` missing `locate` command
2. separate `mainnet` and `testnet` docker images, support upgrading testnet firstly.

zhaojun-sh commented:
@Iruwen Can you review this PR?

Iruwen commented:
For future reference, ath this point there's four possible variants:
1) mainnet image without --testnet argument
2) mainnet image with --testnet argument
3) testnet image without --testnet argument
4) testnet image with --testnet argument
e.g.
1) docker create ... fusionnetwork/efsn
2) docker create ... fusionnetwork/efsn --testnet
3) docker create ... fusionnetwork/testnet-efsn
4) docker create ... fusionnetwork/testnet-efsn --testnet
Setting up a mainnet node from the script equals variant 1) while setting up a testnet node equals variant 4) - variant 2) and 3) aren't covered by the script.

cross-chain commented:
> For future reference, ath this point there's four possible variants:
> 
> 1. mainnet image without --testnet argument, e.g. docker create ... fusionnetwork/efsn
> 2. mainnet image with --testnet argument, e.g. docker create ... fusionnetwork/efsn --testnet
> 3. testnet image without --testnet argument, e.g. docker create ... fusionnetwork/testnet-efsn
> 4. testnet image with --testnet argument, e.g. docker create ... fusionnetwork/testnet-efsn --testnet
> 
> Setting up a mainnet node from the script equals variant 1 while setting up a testnet node equals variant 4 - variant 2 and 3 aren't covered by the script.

@Iruwen Good idea!

At the time we have updated mainnet and testnet docker image,  the docker images for mainnet and testnet are the same one (just different image names, eg. fusionnetwork/efsn:latest, testnet-fusionnetwork/efsn:latest). At this time the variants you mentioned above has variant 1 equal variant 3, and variant 2 equal variant 4.

We did have 4 variants at the time we have updated testnet image but not update the mainnet image. At this time, the mainnet and testnet images are different, but we can not support users to choose which image to run. The user can only use the latest mainnet image for running mainnet efsn node, and can only use the latest testnet image for running testnet efsn node.

So, my suggestion is add a `Please select efsn version` step in `fsnNode.sh`, and select different `tags` according to the user's reply. 

Using tags can support the 4 wariants you mentioned above and more variants of old version image, let's say we have update testnet to v3.6.0, but mainnet is still v3.2.0, in this situation:
>variant 1 is equal to choose latest of mainet image
variant 2 is equal to choose v3.2.0 of testnet image
variant 3 is equal to choose v3.6.0 of mainnet image
variant 4 is equal to choose latest of testnet image

Using tag still have a little difference where variant 3 is not selectable before updating mainnet image as it's not exist. In your situation, we can choose latest testnet image to run a mainnet efsn node. Maybe we can update mainnet image of tag `v3.6.0` firstly to allow user to choose this version, and update tag `latest` when releasing mainnet.

Personally I prefer to separate mainnet and testnet image conceptually, and use tags for old verisons.

Iruwen commented:
I have a few other improvements on my list, back on wednesday then I'll look into it.

Iruwen commented:
I'll use the Docker Hub API to get a list of available tags for each image and allow the user to select which one to install, e.g.
https://registry.hub.docker.com/v2/repositories/fusionnetwork/gateway/tags
I tend to exclude "latest" because it's unreliable (https://vsupalov.com/docker-latest-tag/).
